Dart Target

lingyv-li opened the pull request:
<!--
Thank you for proposing a contribution to the ANTLR project. In order to accept changes from the outside world, all contributors must "sign" the  [contributors.txt](https://github.com/antlr/antlr4/blob/master/contributors.txt) contributors certificate of origin. It's an unfortunate reality of today's fuzzy and bizarre world of open-source ownership.

Make sure you are already in the contributors.txt file or add a commit to this pull request with the appropriate change. Thanks!
-->

This PR introduces Dart target. It passes ~~all existing test cases~~ all test cases except for the large lexer case.

## Why Dart
IMAO Dart is currently the best language that runs easily on browser, Android, iOS, Windows, Linux and MacOS. Also it supports AOT compilation which surprised me by passing `PerformanceTest.DropLoopEntryBranchInLRRule_4` with ease. This test case takes ages to run with JIT+VM, but with AOT, it finishes within two seconds.

## Acknowledgments
This target's runtime is built from scratch by converting the Java target. However in many places I referred to @tiagomazzutti 's [port](https://github.com/tiagomazzutti/antlr4dart-runtime) which helped a lot. Many thanks to @tiagomazzutti and @thosakwe for previous attempts.

Also thank you @marcospassos for adding the codegen tweaks which makes building this target a lot easier.

## TODO

I will add some documentations if the code looks good.

- [x] Document releasing procedure
- [x] Add in `doc/dart-target.md` and `doc/targets.md`
- [x] Add more information to `doc/creating-a-language-target.md` so it's easier for future targets

Also this is my first open source contribution, so please let me know if anything needs to be fixed!

Fixes #1367

parrt commented:
Hiya.  I don't have time to review this at the moment but I have flip to the requested reviewer to @ericvergnaud  to see if he has time. He is the architect of the testing architecture which is the thing we really want to make sure works.

graknol commented:
This looks really interesting.

@lingyv-li 
Quoting @ericvergnaud, what's keeping the larger lexer test from passing?

graknol commented:
Hi,
Got them all to complete.

The issue here was that there's an ongoing bug in `dart2native` on the stable channel.
This caused `dart2native` to fail the AOT compilation, thus resulting in no executable being output. Consequently there was no executable to run the test in question, which explains the empty test output.

When using the beta channel, all the tests complete.
This is a non-blocking issue, because it's not an issue with the target.

If a user uses dart2native, they would already have this issue with their own code.
If the user uses Flutter, it will just work (as Flutter doesn't use dart2native).

ericvergnaud commented:
@parrt this looks good to me, I can't comment on the Dart code itself, but from a testing point of view this is on par with other targets
the impact on the tool is minimal, so suggest we merge this, to be included in a major release

graknol commented:
> @parrt this looks good to me, I can't comment on the Dart code itself, but from a testing point of view this is on par with other targets
> the impact on the tool is minimal, so suggest we merge this, to be included in a major release

Just came to think of the Travis CI build; won't it have to be updated to include the dart-sdk, or does the VM already have it installed?

ericvergnaud commented:
according to this:
https://travis-ci.org/github/antlr/antlr4/builds/641887368

the Dart build passes without any issue


> Le 27 avr. 2020 à 20:18, Sindre Tellevik <notifications@github.com> a écrit :
> 
> 
> @parrt <https://github.com/parrt> this looks good to me, I can't comment on the Dart code itself, but from a testing point of view this is on par with other targets
> the impact on the tool is minimal, so suggest we merge this, to be included in a major release
> 
> Just came to think of the Travis CI build; won't it have to be updated to include the dart-sdk, or does the VM already have it installed?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/antlr/antlr4/pull/2728#issuecomment-619949368>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAZNQJFDX4TVI2QO52HDF5DROVZX3ANCNFSM4KIYSHEQ>.
>

parrt commented:
Hmm...does this look right for testing on CI? Seems like we're missing some templates like `definedexecParser`.

```
Still running (6 of 40): ../.travis/run-tests-dart.sh
The command ../.travis/run-tests-dart.sh exited with 0.
Log:
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
...........................................................Running [Dart:AltNum]
context [anonymous] 6:15 attribute String isn't definedexecParser: /usr/bin/dart Test.dart /tmp/BaseDartTest-pool-1-thread-8-253112526824/input
context [anonymous] 6:15 attribute String isn't definedexecParser: /usr/bin/dart Test.dart /tmp/BaseDartTest-pool-1-thread-6-253113287779/input
context [anonymous] 6:15 attribute String isn't definedexecParser: /usr/bin/dart Test.dart /tmp/BaseDartTest-pool-1-thread-2-253113997080/input
context [anonymous] 6:15 attribute String isn't definedexecParser: /usr/bin/dart Test.dart /tmp/BaseDartTest-pool-1-thread-7-253125351671/input
context [anonymous] 6:15 attribute String isn't definedexecParser: /usr/bin/dart Test.dart /tmp/BaseDartTest-pool-1-thread-1-253125351646/input
...
```

ericvergnaud commented:
Indeed, sorry I overlooked, assuming the test results were reflected.

> Le 28 avr. 2020 à 00:07, Terence Parr <notifications@github.com> a écrit :
> 
> 
> Hmm...does this look right for testing on CI? Seems like we're missing some templates like definedexecParser.
> 
> Still running (6 of 40): ../.travis/run-tests-dart.sh
> The command ../.travis/run-tests-dart.sh exited with 0.
> Log:
> -------------------------------------------------------
>  T E S T S
> -------------------------------------------------------
> ...........................................................Running [Dart:AltNum]
> context [anonymous] 6:15 attribute String isn't definedexecParser: /usr/bin/dart Test.dart /tmp/BaseDartTest-pool-1-thread-8-253112526824/input
> context [anonymous] 6:15 attribute String isn't definedexecParser: /usr/bin/dart Test.dart /tmp/BaseDartTest-pool-1-thread-6-253113287779/input
> context [anonymous] 6:15 attribute String isn't definedexecParser: /usr/bin/dart Test.dart /tmp/BaseDartTest-pool-1-thread-2-253113997080/input
> context [anonymous] 6:15 attribute String isn't definedexecParser: /usr/bin/dart Test.dart /tmp/BaseDartTest-pool-1-thread-7-253125351671/input
> context [anonymous] 6:15 attribute String isn't definedexecParser: /usr/bin/dart Test.dart /tmp/BaseDartTest-pool-1-thread-1-253125351646/input
> ...
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/antlr/antlr4/pull/2728#issuecomment-620081323>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAZNQJGZ762U3BWK4FPHJE3ROWUTFANCNFSM4KIYSHEQ>.
>

graknol commented:
@ericvergnaud
What do those `context [anonymous] 6:15 attribute String isn't definedexecParser` messages mean?

Should it report them as successful if they actually fail?

parrt commented:
@graknol that's a StringTemplate (ST4) message indicating `String` (a weird name) is referenced from a template called `execParser` I think.

graknol commented:
> @graknol that's a StringTemplate (ST4) message indicating `String` (a weird name) is referenced from a template called `execParser` I think.

The issue was that some generics were not escaped (`<String>` --> `\<String>` and `<DFA>` --> `\<DFA>`).
After making these changes, it's stopped complaining.

Also removed the debug messages `execParser: /usr/bin/dart Test.dart /tmp/BaseDartTest...`

After running all dart tests we get:
```log
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.antlr.v4.test.runtime.dart.TestCompositeLexers
.Not skippingC:\Users\SC-23431\AppData\Local\Temp\BaseDartTest-main-429499808126300

Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.042 sec - in org.antlr.v4.test.runtime.dart.TestCompositeLexers
Running org.antlr.v4.test.runtime.dart.TestCompositeParsers
...Tests run: 15, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 13.323 sec - in org.antlr.v4.test.runtime.dart.TestCompositeParsers
Running org.antlr.v4.test.runtime.dart.TestFullContextParsing
..Tests run: 15, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 13.529 sec - in org.antlr.v4.test.runtime.dart.TestFullContextParsing
Running org.antlr.v4.test.runtime.dart.TestLeftRecursion
..................Tests run: 98, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 87.684 sec - in org.antlr.v4.test.runtime.dart.TestLeftRecursion
Running org.antlr.v4.test.runtime.dart.TestLexerErrors
..Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 9.699 sec - in org.antlr.v4.test.runtime.dart.TestLexerErrors
Running org.antlr.v4.test.runtime.dart.TestLexerExec
Compile: C:\tools\dart-sdk\bin\dart2native.bat Test.dart -o Test
...........Tests run: 37, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 57.572 sec - in org.antlr.v4.test.runtime.dart.TestLexerExec
Running org.antlr.v4.test.runtime.dart.TestListeners
..Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 6.363 sec - in org.antlr.v4.test.runtime.dart.TestListeners
Running org.antlr.v4.test.runtime.dart.TestParserErrors
.....Tests run: 34, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 28.44 sec - in org.antlr.v4.test.runtime.dart.TestParserErrors
Running org.antlr.v4.test.runtime.dart.TestParserExec
.......Tests run: 37, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 30.913 sec - in org.antlr.v4.test.runtime.dart.TestParserExec
Running org.antlr.v4.test.runtime.dart.TestParseTrees
.Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 8.278 sec - in org.antlr.v4.test.runtime.dart.TestParseTrees
Running org.antlr.v4.test.runtime.dart.TestPerformance
..Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 7.878 sec - in org.antlr.v4.test.runtime.dart.TestPerformance
Running org.antlr.v4.test.runtime.dart.TestSemPredEvalLexer
.Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 6.389 sec - in org.antlr.v4.test.runtime.dart.TestSemPredEvalLexer
Running org.antlr.v4.test.runtime.dart.TestSemPredEvalParser
....Tests run: 26, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 21.185 sec - in org.antlr.v4.test.runtime.dart.TestSemPredEvalParser
Running org.antlr.v4.test.runtime.dart.TestSets
......Tests run: 32, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 26.665 sec - in org.antlr.v4.test.runtime.dart.TestSets

Results :

Tests run: 340, Failures: 0, Errors: 0, Skipped: 1

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  05:25 min
[INFO] Finished at: 2020-04-28T07:21:20+02:00
[INFO] ------------------------------------------------------------------------
```

I couldn't seem to figure out why it skipped one of the tests in `TestSemPredEvalParser`, but perhaps that is the case for every target?

ericvergnaud commented:
> I couldn't seem to figure out why it skipped one of the tests in `TestSemPredEvalParser`, but perhaps that is the case for every target?

It's due to the following:	
```
public boolean ignore(String targetName) {
    return !"Java".equals(targetName) && !"Swift".equals(targetName);
}
```

Great that you got the tests to pass !
I'll let @parrt opine on the next steps

parrt commented:
Seems like a good sign that they pass but I'm worried that we got a weird message that seemed important but they passed.  @graknol have you used it for parsing real stuff like Java input?

graknol commented:
@parrt
We got the message to disappear (needed to escape some "<")
I have actually not generated any parser with it yet. I just popped in to see if I could help, but the tests do run and parse stuff. I'm pretty busy with my finals and thesis, but perhaps @lingyv-li has done so?

parrt commented:
Yeah, we really need more verification before we can merge this in.

mqus commented:
Hey! I would really like to use it to generate a parser for SQLite queries. I'm coming from https://pub.dev/packages/antlr4 where nothing is documented, and read your comments.
So how do I actually integrate this in my dart/flutter workflow or even generate a parser from an antlr grammar? I would be using the [updated one by jdkoren](https://github.com/jdkoren/sqlite-parser/blob/v3.24/src/main/antlr4/nl/bigo/sqliteparser/SQLite.g4)

(Just mentioning, I haven't worked with antlr yet and my last encounter with parser generators is a few years back but I'm willing to put in some effort :wink: )

mqus commented:
So I wrote a small program to parse a SQLite Query using the generated dart sources. I have a few other minor issues but more importantly there seem to be errors in the generated Code in the SQLiteParser.dart file:
```dart

  ParseContext parse() {
    dynamic _localctx = new ParseContext(context, state);
    enterRule(_localctx, 0, RULE_parse);
    int _la;
    try {
      enterOuterAlt(_localctx, 1);
      {
      state = 172;
      errorHandler.sync(this);
      _la = tokenStream.LA(1);
      while ((((_la) & ~0x3f) == 0 && ((BigInt.one << _la) & ((BigInt.one << TOKEN_SCOL) | (BigInt.one << TOKEN_K_ALTER) | (BigInt.one << TOKEN_K_ANALYZE) | (BigInt.one << TOKEN_K_ATTACH) | (BigInt.one << TOKEN_K_BEGIN) | (BigInt.one << TOKEN_K_COMMIT) | (BigInt.one << TOKEN_K_CREATE) | (BigInt.one << TOKEN_K_DELETE) | (BigInt.one << TOKEN_K_DETACH))) != BigInt.zero) || ((((_la - 65)) & ~0x3f) == 0 && ((BigInt.one << (_la - 65)) & ((BigInt.one << TOKEN_(K_DROP - 65)) | (BigInt.one << TOKEN_(K_END - 65)) | (BigInt.one << TOKEN_(K_EXPLAIN - 65)) | (BigInt.one << TOKEN_(K_INSERT - 65)) | (BigInt.one << TOKEN_(K_PRAGMA - 65)) | (BigInt.one << TOKEN_(K_REINDEX - 65)) | (BigInt.one << TOKEN_(K_RELEASE - 65)) | (BigInt.one << TOKEN_(K_REPLACE - 65)) | (BigInt.one << TOKEN_(K_ROLLBACK - 65)))) != BigInt.zero) || ((((_la - 129)) & ~0x3f) == 0 && ((BigInt.one << (_la - 129)) & ((BigInt.one << TOKEN_(K_SAVEPOINT - 129)) | (BigInt.one << TOKEN_(K_SELECT - 129)) | (BigInt.one << TOKEN_(K_UPDATE - 129)) | (BigInt.one << TOKEN_(K_VACUUM - 129)) | (BigInt.one << TOKEN_(K_VALUES - 129)) | (BigInt.one << TOKEN_(K_WITH - 129)) | (BigInt.one << TOKEN_(UNEXPECTED_CHAR - 129)))) != BigInt.zero)) {
        {
        state = 170;
        errorHandler.sync(this);
        switch (tokenStream.LA(1)) {
        case TOKEN_SCOL:
        case TOKEN_K_ALTER:
        case TOKEN_K_ANALYZE:
        case TOKEN_K_ATTACH:
        case TOKEN_K_BEGIN:
        case TOKEN_K_COMMIT:
        case TOKEN_K_CREATE:
        case TOKEN_K_DELETE:
        case TOKEN_K_DETACH:
        case TOKEN_K_DROP:
        case TOKEN_K_END:
        case TOKEN_K_EXPLAIN:
        case TOKEN_K_INSERT:
        case TOKEN_K_PRAGMA:
        case TOKEN_K_REINDEX:
        case TOKEN_K_RELEASE:
        case TOKEN_K_REPLACE:
        case TOKEN_K_ROLLBACK:
        case TOKEN_K_SAVEPOINT:
        case TOKEN_K_SELECT:
        case TOKEN_K_UPDATE:
        case TOKEN_K_VACUUM:
        case TOKEN_K_VALUES:
        case TOKEN_K_WITH:
          {
          state = 168;
          sql_stmt_list();
          }
          break;
        case TOKEN_UNEXPECTED_CHAR:
          {
          state = 169;
          error();
          }
          break;
        default:
          throw new NoViableAltException(this);
        }
        }
        state = 174;
        errorHandler.sync(this);
        _la = tokenStream.LA(1);
      }
      state = 175;
      match(TOKEN_EOF);
      }
    } on RecognitionException catch (re) {
      _localctx.exception = re;
      errorHandler.reportError(this, re);
      errorHandler.recover(this, re);
    }  finally {
      exitRule();
    }
    return _localctx;
  }
```
In the (long) while condition there are later many clauses where it suddenly generates snippets like the following:
```dart
(BigInt.one << TOKEN_(K_VALUES - 129)) | (BigInt.one << TOKEN_(K_WITH - 129))
```
The opening bracket within the '`TOKEN_K_VALUES`' seems to be wrong. The dart compiler complains about an unknown function `TOKEN_(...)` and an unknown field/getter `K_VALUES`,`K_WITH`, etcpp. But after running regex-replacing `TOKEN_\(([A-Z_]+)` with `(TOKEN_$1` everything compiled.

The smaller issues are mostly about documentation and dart patterns. I'll summarize them after I did a bit more work.

graknol commented:
@mqus
At the end of line 536 in the file tool/.../templates/Dart/Dart.stg you'll find this:
```st
TOKEN_<offsetShift(ttype, bits.shift)>
```
I don't know StringTemplate that well, but you need to be:
```st
<offsetShift('TOKEN_' + ttype, bits.shift)>
```
(This is because the offsetShift partial below wraps everything in parenthesis.)

*I'm currently on the phone rn, on a ferry, so I can't make the change myself.*

mqus commented:
I tried that just now but that seems to result in `TOKEN_<ttype>` in the generated code and not the intended `TOKEN_K_VALUES`. I tried different things but as I know nothing about the stg format I didn't get anything to work.

mqus commented:
I added a simple addTokenPrefix "function"(?) which seems to work for now. 
```stg
bitsetBitfieldComparison(s, bits) ::= <%
(<testShiftInRange({<offsetShift(s.varName, bits.shift)>})> && ((BigInt.one \<\< <offsetShift(s.varName, bits.shift)>) & (<bits.ttypes:{ttype | (BigInt.one \<\< <offsetShift(addTokenPrefix(ttype), bits.shift)>)}; separator=" | ">)) != BigInt.zero)
%>

addTokenPrefix(s) ::= "TOKEN_<s>"
```
I think it doesn't make sense to open another PR on this or the other branch for such a small change, or should I do it anyway?

ericvergnaud commented:
Hi guys,

thanks so much for your involvement

it seems the existing test suite does not cover these productions
whilst you’re at it, it would make sense to add them

the way to do that is to add a static inner class in an existing test descriptor
the best landing place for this one could be SetsDescriptors

the trick is to identify the smallest possible grammar fragment that triggers the failing production

would you mind doing that?
ideally the CI would fail before making the below change and pass after implementing it

Cheers,

Eric

> Le 3 mai 2020 à 01:53, mqus <notifications@github.com> a écrit :
> 
> 
> I added a simple addTokenPrefix "function"(?) which seems to work for now.
> 
> bitsetBitfieldComparison(s, bits) ::= <%
> (<testShiftInRange({<offsetShift(s.varName, bits.shift)>})> && ((BigInt.one \<\< <offsetShift(s.varName, bits.shift)>) & (<bits.ttypes:{ttype | (BigInt.one \<\< <offsetShift(addTokenPrefix(ttype), bits.shift)>)}; separator=" | ">)) != BigInt.zero)
> %>
> 
> addTokenPrefix(s) ::= "TOKEN_<s>"
> I think it doesn't make sense to open another PR on this or the other branch for such a small change, or should I do it anyway?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/antlr/antlr4/pull/2728#issuecomment-622990855>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAZNQJF6K7ZO2RD5365SURDRPRM27ANCNFSM4KIYSHEQ>.
>

mqus commented:
I know way too little about the internals (and externals!) of antlr to confidently add tests for that usecase. From what I understand the error occurs because there are >64 Keywords (130+) in the grammar and it needs to shift those constants to fit inside the bitmap. But everything else leading to that bitmap (and the amount that is subtracted) is pure magic to me.

graknol commented:
@ericvergnaud So we should write the smallest possible parser grammar which produces the wrong Dart code?

This should then ignore all other targets, right?

It would also be marked as a regression test then?

ericvergnaud commented:
Hi,

yes you should write the smallest possible parser grammar which produces the wrong Dart code
all tests are regression tests… and all targets should pass this new test
(the only ignored test is the ‘big’ performance test which takes ages with slow runtimes (python, javascript...))

the philosophy here is that such a test is missing hence your CI is green, despite a ‘hole’ in the target template
thanks to the test we may detect similar issues in other targets (obviously for their authors to fix)
and if some later change breaks the behavior we will know immediately from the CI

does that make sense?

Eric

> Le 3 mai 2020 à 22:13, Sindre Tellevik <notifications@github.com> a écrit :
> 
> 
> @ericvergnaud <https://github.com/ericvergnaud> So we should write the smallest possible parser grammar which produces the wrong Dart code?
> 
> This should then ignore all other targets, right?
> 
> It would also be marked as a regression test then?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/antlr/antlr4/pull/2728#issuecomment-623116297>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAZNQJC6VADPVTCYVEE7HFDRPV3YPANCNFSM4KIYSHEQ>.
>

lingyv-li commented:
Sorry for being inactive. I've set GitHub to my old email. 😢 

@mqus  I fixed the issue and added a regression test, thank you for finding it. The implementation here is trying to be as close to the Java implementation as possible for maintenance, but I'd love to know your thoughts on dart patterns!

I have merged in @graknol 's fix (thank you!) and will add docs later this week. Please let me know on any other issues.

mqus commented:
I'll look at it more tomorrow (and find out what I meant when I wrote that :D) but one thing that I noticed just now:
The generated parser contains
```dart
  static final void _versionCheck = () {
    RuntimeMetaData.checkVersion("4.8", RuntimeMetaData.VERSION);
  }();
```
I don't think that this gets run when initalizing the parser (as you probably intended). It gets only run when accessing `_versionCheck`, which does not happen in the generated code.

mqus commented:
Ok, most of the 'dart-patterns' are things which dart/flutter analyzer complains about.
Some examples:
- You don't need to (and should not) use a `new` in front of constructors.
- use final everywhere(you can), use dynamic only if you must. In some cases I saw you can replace dynamic with final without losing anything.
- String literals should use `'` delimiters and not `"` if the string does not contain `'`
- the line in the previous comment was actually altered by me, the original line was 
```dart
static final _versionCheck = () { RuntimeMetaData.checkVersion("4.8", RuntimeMetaData.VERSION); }();
```
where dart (rightfully) complains that the type information is missing

I also mentioned documentation. I originally found this PR by finding the antlr4 package on the dart package repositories (https://pub.dev/packages/antlr4) where the package is already there (and says nothing about being a work in progress), but without content and the only linked documentation is to a dead link you removed in an earlier commit (nevermind that the doc was not in the `master` branch). *All of this* was a really bad experience for me (and definitely also for others who just didn't bother to look further) which was not necessary at all. please don't publish a dart package before it is usable.

Regarding the docs: It would be nice if the dart-target.md file would contain a section on how to include the parser building workflow into the dart toolchain. I didn't find out how to do this and decided to just build them once and put them into git. 
Because you altered the java runtime and it maps pretty well into rust the documentation there is pretty good. The only issue is that dart tooling(IDEs, etc) will not be able to work with it because dart docstrings start with `///` in each line and not the javas `/**`,` * `,` */`, see also https://dart.dev/guides/language/effective-dart/documentation regarding that

lingyv-li commented:
@mqus Thank you for your feedback! It is my mistake publishing the package. I have marked the package as discontinued earlier this week after seeing you are from the package. Sorry for the inconvenience.

> I don't think that this gets run when initalizing the parser (as you probably intended). It gets only run when accessing `_versionCheck`, which does not happen in the generated code.

Yes, the version check is not executed, and there is no way to make it run only once automatically when start up. I'll make this method public allowing users to call it manually on start up.

> Ok, most of the 'dart-patterns' are things which dart/flutter analyzer complains about.
> Some examples:
> 
> You don't need to (and should not) use a new in front of constructors.
> use final everywhere(you can), use dynamic only if you must. In some cases I saw you can replace dynamic with final without losing anything.
> String literals should use ' delimiters and not " if the string does not contain '

I'll run the code analyser against google's pedantic package, and that should cover these and other guidelines in Effective Dart.

> Regarding the docs: It would be nice if the dart-target.md file would be reinstated and contain a section on how to include the parser building workflow into the dart toolchain. I didn't find out how to do this and decided to just build them once and put them into git.

Docs will come later, but I also cannot think of an easy way to integrate it with the dart toolchain yet. I would do codegen in a separate repo with CI/CD pipelines. Thoughts and contributions welcomed.

mqus commented:
It does seem like code generation or other tools can't be integrated in the `flutter test` or similar commands easily... I'm working on a package which does code generation myself but you always have to `pub run build-runner xyz`... I think if it's hard then it can be added later as well ;)

lingyv-li commented:
Added docs and fixed code analyzer warnings.
For publishing, looks like we need to have a verified publisher https://pub.dev/create-publisher to transfer package ownership. I can transfer ownership after having a publisher.

@ericvergnaud can you please review this PR again? Thank you!

ericvergnaud commented:
@lingyv-li as far as the code concerned, it seems ok in the sense that it does not affect existing code
the tests now seem to pass without error:
Results :
Tests run: 341, Failures: 0, Errors: 0, Skipped: 1

but looks like there is still room for improvement re the build quality ? The below looks like an error but is not reported (the build is green) ?

/home/travis/.travis/functions: line 607:  5363 Terminated              travis_jigger "${!}" "${timeout}" "${cmd[@]}"
cat: 'target/surefire-reports/*.dumpstream': No such file or directory
 
also do you have evidence that the runtime runs successfully on a complex grammar?

this is great, really!
but the moment this is incorporated, it needs to be supported accurately, we can't afford to release versions of antlr4 with a target deemed ok but not actually ok.

lingyv-li commented:
Thank you @ericvergnaud 

I think the *.dumpstream error is the same as [this one](https://github.com/antlr/antlr4/issues/2698#issuecomment-563342208). It means there is no crash dump.

I don't have it running on a complex grammar and may need help doing it. If there is an example on Java I should be able to do the same.

ericvergnaud commented:
> I think the *.dumpstream error is the same as [this one](https://github.com/antlr/antlr4/issues/2698#issuecomment-563342208). It means there is no crash dump.
ok, then can you introduce an error such that the build fails? We just want to prove that if a test fails the travoise build goes red

> I don't have it running on a complex grammar and may need help doing it. If there is an example on Java I should be able to do the same.
You could try https://github.com/antlr/grammars-v4/tree/master/java/java9
Simply replace Java inline code by corresponding Dart code, and run that against a Java source code (Terence runs that grammar using the Java runtime against the entire Java SDK)

lingyv-li commented:
> ok, then can you introduce an error such that the build fails? We just want to prove that if a test fails the travoise build goes red

Done. Never thought that needs proofing.
https://travis-ci.org/github/antlr/antlr4/jobs/695402499

> You could try https://github.com/antlr/grammars-v4/tree/master/java/java9
Simply replace Java inline code by corresponding Dart code, and run that against a Java source code

See this repo: https://github.com/lingyv-li/java9parser

The Java inline code is calling some JDK functions (`isJavaIdentifierStart`, `isJavaIdentifierPart`) so I simplified it to `true`.

Here is what happens when I ran against the example files in https://github.com/antlr/grammars-v4/tree/master/java/java9/examples

Plus an additional invalid java file `invalidhelloworld.java`.
```sh
(base) ➜  java9parser git:(master) ✗ dart bin/main.dart examples/IdentifierTest.java   
(base) ➜  java9parser git:(master) ✗ dart bin/main.dart examples/module-info.java   
(base) ➜  java9parser git:(master) ✗ dart bin/main.dart examples/TryWithResourceDemo.java 
(base) ➜  java9parser git:(master) ✗ dart bin/main.dart examples/invalidhelloworld.java  
line 3:41 no viable alternative at input 'System.out.println("Hello, World";'
(base) ➜  java9parser git:(master) ✗ 
```
If you feel necessary I can run it against the Java SDK as well.

ericvergnaud commented:
Ok, thanks for the Travis proof. Closing that bit.

Re java parsing, seems looking good!

@parrt wanna have a second look?

zekunyan commented:
We also need Dart version ANTLR. When will this release ?

modulovalue commented:
Hey @parrt, this is so close to being done, could you perhaps take a look when you get a chance. Thank you!

parrt commented:
Hiya. Yeah, i'm trying to decide how I'll know if the DART target is mature enough to merge.

zekunyan commented:
Really looking forward to using it~~!

parrt commented:
hi. What grammars / translators etc... has this been tried with?

lingyv-li commented:
Those grammars in the unit tests and the Java grammar in the previous comments. Please let me know what else grammars it needs to be tested with.

parrt commented:
Well, Java grammar is good one but testing on more from https://github.com/antlr/grammars-v4 would make sense.

lingyv-li commented:
@parrt Many grammars in that repo either use Dart keywords (`var`, `operator`) or have language specific code block. It'll be very time consuming to clean up each grammar and test. Can you please be more specific which grammars you would like to see testing?

parrt commented:
I thought we had a mechanism for each target that specified the keywords that get modified during translation so you should not have to worry about this

lingyv-li commented:
There is a mechanism to catch that and raise an error. Can you point me to the code that modify the keywords?
https://github.com/antlr/antlr4/blob/eb1adaa8a7a6993803f08f5ffa0350a45eb24a13/tool/src/org/antlr/v4/codegen/CodeGenPipeline.java#L35
https://github.com/antlr/antlr4/blob/eb1adaa8a7a6993803f08f5ffa0350a45eb24a13/tool/src/org/antlr/v4/codegen/target/JavaTarget.java#L74

lingyv-li commented:
@parrt May I ask, taking PHP Target #2645 as an example, what is making it mature enough to merge? So I can do the same.

parrt commented on appveyor.yml:
```diff
@@ -8,14 +8,15 @@ install:
 after_build: 
   - msbuild /target:pack /property:Configuration=Release /verbosity:detailed runtime/CSharp/runtime/CSharp/Antlr4.dotnet.sln
 test_script:
-  - mvn install -Dantlr-php-php="C:\tools\php\php.exe" -Dantlr-python2-python="C:\Python27\python.exe" -Dantlr-python3-python="C:\Python35\python.exe" -Dantlr-javascript-nodejs="C:\Program Files (x86)\nodejs\node.exe" --batch-mode
```
is this deleting the test for php?  shouldn't you be adding something and not deleting?

lingyv-li replied:
Not at all. The change on this line is adding `dart`, `dart pub` and `dart2native` executable mapping.

parrt replied:
oh sorry. yeah it repeated the php part. cool.

parrt commented:
Hi.  I guess we can call it preliminary. Started looking through and making a few comments

parrt commented:
ok,  I just had that one comment. it looks pretty complete!  thank you for the hard work!

@ericvergnaud  do you have any unresolved comments in the code?

parrt commented:
Ok, merging!

parrt commented:
Thanks for all the hard work!

lingyv-li commented:
Thank you! Can you please also create a publisher account so that I can transfer the package ownership?
https://pub.dev/create-publisher

parrt commented:
I created an account with parrt@cs.usfca.edu but i'm not sure i can do the whole publisher thing. is that necessary?

lingyv-li commented:
To create a publisher you'll need to verify a domain name with Google (following the instruction). I can use my personal domain and add you as an admin if you can't do that, but that's less ideal.

parrt commented:
Yeah, i don't wanna change my domain records. what do i need the publisher thing for?

lingyv-li commented:
It is for publishing the runtime to pub.dev, the package manager of Dart. It will be needed when upgrading the runtime version by doing `pub publish`, and it will also show up on the package homepage https://pub.dev/packages/antlr4

parrt commented:
Hmm...i wonder why it's that complicated. Well, i don't wanna mess with my domain records, sorry!

lingyv-li commented:
No worries. I'll use my domain then, thanks!
